model/stats/query.go

```diff
@@ -0,0 +1,284 @@
+package stats
+
+import (
+	"time"
+
+	"github.com/evergreen-ci/evergreen/db"
+	"github.com/evergreen-ci/evergreen/util"
+	"github.com/mongodb/grip"
+	"github.com/pkg/errors"
+)
+
+const (
+	MaxQueryLimit             = 1000
+	GroupByTest       GroupBy = "test"
+	GroupByTask       GroupBy = "task"
+	GroupByVariant    GroupBy = "variant"
+	GroupByDistro     GroupBy = "distro"
+	SortEarliestFirst Sort    = "earliest"
+	SortLatestFirst   Sort    = "latest"
+)
+
+type GroupBy string
+
+func (gb *GroupBy) validate() error {
+	switch *gb {
+	case GroupByDistro:
+	case GroupByVariant:
+	case GroupByTask:
+	case GroupByTest:
+	default:
+		return errors.Errorf("Invalid GroupBy value: %v", *gb)
+	}
+	return nil
+}
+
+type Sort string
+
+func (s *Sort) validate() error {
+	switch *s {
+	case SortLatestFirst:
+	case SortEarliestFirst:
+	default:
+		return errors.Errorf("Invalid Sort value: %v", *s)
+	}
+	return nil
+}
+
+/////////////
+// Filters //
+/////////////
+
+// Represents parameters that allow a search query to resume at a specific point.
+// Used for pagination.
+type StartAt struct {
+	Date         time.Time
+	BuildVariant string
+	Task         string
+	Test         string
+	Distro       string
+}
+
+// Validates that the StartAt struct is valid for use with test stats.
+func (s *StartAt) validateCommon(groupBy GroupBy) error {
+	catcher := grip.NewBasicCatcher()
+	if !s.Date.Equal(util.GetUTCDay(s.Date)) {
+		catcher.Add(errors.New("Invalid StartAt Date value"))
+	}
+	switch groupBy {
+	case GroupByDistro:
+		if len(s.Distro) == 0 {
+			catcher.Add(errors.New("Missing StartAt Distro value"))
+		}
+		fallthrough
+	case GroupByVariant:
+		if len(s.BuildVariant) == 0 {
+			catcher.Add(errors.New("Missing StartAt BuildVariant value"))
+		}
+		fallthrough
+	case GroupByTask:
+		if len(s.Task) == 0 {
+			catcher.Add(errors.New("Missing StartAt Task value"))
+		}
+	}
+	return catcher.Resolve()
+}
+
+// Validates that the StartAt struct is valid for use with test stats.
+func (s *StartAt) validateForTests(groupBy GroupBy) error {
+	catcher := grip.NewBasicCatcher()
+	catcher.Add(s.validateCommon(groupBy))
+	if len(s.Test) == 0 {
+		catcher.Add(errors.New("Missing Start Test value"))
+	}
+	return catcher.Resolve()
+}
+
+// Validates that the StartAt struct is valid for use with task stats.
+func (s *StartAt) validateForTasks(groupBy GroupBy) error {
+	catcher := grip.NewBasicCatcher()
+	catcher.Add(s.validateCommon(groupBy))
+	if len(s.Test) != 0 {
+		catcher.Add(errors.New("StartAt for task stats should not have a Test value"))
+	}
+	return catcher.Resolve()
+}
+
+// Represents search and aggregation parameters when querying the test or task statistics.
```

Comment:
Comment style is especially important for public types.

Reply:
Done.